README.md
dalek-browser-ios
=====================

> DalekJS browser plugin for mobile Safario on iOS

[![Build Status](https://travis-ci.org/dalekjs/dalek-browser-ios.png)](https://travis-ci.org/dalekjs/dalek-browser-ios)
[![Build Status](https://drone.io/github.com/dalekjs/dalek-browser-ios/status.png)](https://drone.io/github.com/dalekjs/dalek-browser-ios/latest)
[![Dependency Status](https://david-dm.org/dalekjs/dalek-browser-ios.png)](https://david-dm.org/dalekjs/dalek-browser-ios)
[![devDependency Status](https://david-dm.org/dalekjs/dalek-browser-ios/dev-status.png)](https://david-dm.org/dalekjs/dalek-browser-ios#info=devDependencies)
[![NPM version](https://badge.fury.io/js/dalek-browser-ios.png)](http://badge.fury.io/js/dalek-browser-ios)
[![Coverage](http://dalekjs.com/package/dalek-browser-ios/master/coverage/coverage.png)](http://dalekjs.com/package/dalek-browser-ios/master/coverage/index.html)
[![unstable](https://rawgithub.com/hughsk/stability-badges/master/dist/unstable.svg)](http://github.com/hughsk/stability-badges)
[![Stories in Ready](https://badge.waffle.io/dalekjs/dalek-browser-ios.png?label=ready)](https://waffle.io/dalekjs/dalek-browser-ios)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/dalekjs/dalek-browser-ios/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
[![Built with Grunt](https://cdn.gruntjs.com/builtwith.png)](http://gruntjs.com/)

[![NPM](https://nodei.co/npm/dalek-browser-ios.png)](https://nodei.co/npm/dalek-browser-ios/)
[![NPM](https://nodei.co/npm-dl/dalek-browser-ios.png)](https://nodei.co/npm/dalek-browser-ios/)

## Ressources

[API Docs](http://dalekjs.com/package/dalek-browser-ios/master/api/index.html) -
[Trello](https://trello.com/b/xhw6Jv7A/dalek-browser-ios) -
[Code coverage](http://dalekjs.com/package/dalek-browser-ios/master/coverage/index.html) -
[Code complexity](http://dalekjs.com/package/dalek-browser-ios/master/complexity/index.html) -
[Contributing](https://github.com/dalekjs/dalek-browser-ios/blob/master/CONTRIBUTING.md) -
[User Docs](http://dalekjs.com/docs/chrome.html) -
[Homepage](http://dalekjs.com) -
[Twitter](http://twitter.com/dalekjs)

## Docs

This module is a browser plugin for [DalekJS](//github.com/dalekjs/dalek).
It provides all a WebDriverServer & browser launcher for mobile Safari on iOS devices & the iOS Simulator.

The browser plugin can be installed with the following command:

```
$ npm install dalek-browser-ios --save-dev
```

You can use the browser plugin by adding a config option to the your Dalekfile

```js
"browsers": ["ios:safari"]
```

Or you can tell Dalek that it should test in this browser via the command line:

```
$ dalek mytest.js -b ios:safari
```

## Help Is Just A Click Away

### #dalekjs on FreeNode.net IRC

Join the `#daleksjs` channel on [FreeNode.net](http://freenode.net) to ask questions and get help.

### [Google Group Mailing List](https://groups.google.com/forum/#!forum/dalekjs)

Get announcements for new releases, share your projects and ideas that are
using DalekJS, and join in open-ended discussion that does not fit in
to the Github issues list or StackOverflow Q&A.

**For help with syntax, specific questions on how to implement a feature
using DalekJS, and other Q&A items, use StackOverflow.**

### [StackOverflow](http://stackoverflow.com/questions/tagged/dalekjs)

Ask questions about using DalekJS in specific scenarios, with
specific features. For example, help with syntax, understanding how a feature works and
how to override that feature, browser specific problems and so on.

Questions on StackOverflow often turn in to blog posts or issues.

### [Github Issues](//github.com/dalekjs/dalek-browser-ios/issues)

Report issues with DalekJS, submit pull requests to fix problems, or to
create summarized and documented feature requests (preferably with pull
requests that implement the feature).

**Please don't ask questions or seek help in the issues list.** There are
other, better channels for seeking assistance, like StackOverflow and the
Google Groups mailing list.

![DalekJS](https://raw.github.com/dalekjs/dalekjs.com/master/img/logo.jpg)

## Legal FooBar (MIT License)

Copyright (c) 2013 Sebastian Golasch

Distributed under [MIT license](https://github.com/dalekjs/dalek-browser-ios/blob/master/LICENSE-MIT)

CHANGELOG.md
<a name="0.0.3"></a>
### 0.0.3 (2014-11-28)


#### Bug Fixes

* **core:** added changes needed to use newer appium versions ([df50fdee](http://github.com/dalekjs/dalek-browser-ios/commit/df50fdeefc566aa9859383d898531e45764fc6a1))




test/index_TEST.js
'use strict';

var expect = require('chai').expect;
var IosDriver = require('../index');

describe('dalek-browser-ios', function() {

  it('should get default webdriver port', function(){
    expect(IosDriver.port).to.equal(4723);
  });

  it('should get default webhook port', function(){
    expect(IosDriver.webhookPort).to.equal(9003);
  });

});


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  // check task runtime
  require('time-grunt')(grunt);

  // load generic configs
  var configs = require('dalek-build-tools');

  // project config
  grunt.initConfig({

    // load module meta data
    pkg: grunt.file.readJSON('package.json'),

    // define a src set of files for other tasks
    src: {
      lint: ['Gruntfile.js', 'index.js', 'test/*.js'],
      complexity: ['index.js'],
      test: ['test/*.js'],
      src: ['index.js']
    },
    
    // clean automatically generated helper files & docs
    clean: configs.clean,

    // speed up build by defining concurrent tasks
    concurrent: configs.concurrent,

    // linting
    jshint: configs.jshint,

    // testing
    mochaTest: configs.mocha,

    // code metrics
    complexity: configs.complexity,
    plato: configs.plato(grunt.file.readJSON('.jshintrc')),

    // api docs
    yuidoc: configs.yuidocs(),

    // user docs
    documantix: {
      options: {
        header: 'dalekjs/dalekjs.com/master/assets/header.html',
        footer: 'dalekjs/dalekjs.com/master/assets/footer.html',
        target: 'report/docs',
        vars: {
          title: 'DalekJS - Documentation - Browser - Mobile Safari on iOS',
          desc: 'DalekJS - Documentation - Browser - Mobile Safari on iOS',
          docs: true
        }
      },
      src: ['index.js']
    },

    // add current timestamp to the html document
    includereplace: {
      dist: {
        options: {
          globals: {
            timestamp: '<%= grunt.template.today("dddd, mmmm dS, yyyy, h:MM:ss TT") %>'
          },
        },
        src: 'report/docs/*.html',
        dest: './'
      }
    },

    // up version, tag & commit
    bump: configs.bump({
      pushTo: 'git@github.com:dalekjs/dalek-browser-ios.git',
      files: ['package.json', 'CONTRIBUTORS.md', 'CHANGELOG.md']
    }),

    // generate contributors file
    contributors: configs.contributors,

    // compress artifacts
    compress: configs.compress,

    // prepare files for grunt-plato to
    // avoid error messages (weird issue...)
    preparePlato: {
      options: {
        folders: [
          'coverage',
          'report',
          'report/coverage',
          'report/complexity',
          'report/complexity/files',
          'report/complexity/files/index_js'
        ],
        files: [
          'report.history.json',
          'files/index_js/report.history.json'
        ]
      }
    },

    // prepare files & folders for coverage
    prepareCoverage: {
      options: {
        folders: ['coverage', 'report', 'report/coverage'],
        pattern: '[require("fs").realpathSync(__dirname + "/../index.js")]'
      }
    },

    // list requires that need to be changed
    // for generating a canary build
    'release-canary': {
      options: {
        files: ['index.js']
      }
    },

    // archives the docs if a new version appears
    archive: {
      options: {
        file: 'ios.html'
      }
    }

  });

  // load 3rd party tasks
  require('load-grunt-tasks')(grunt);
  grunt.loadTasks('./node_modules/dalek-build-tools/tasks');

  // define runner tasks
  grunt.registerTask('lint', 'jshint');
  
  // split test & docs for speed
  grunt.registerTask('test', ['clean:coverage', 'prepareCoverage', 'concurrent:test', 'generateCoverageBadge']);
  grunt.registerTask('docs', ['clean:reportZip', 'clean:report', 'preparePlato', 'documantix', 'includereplace', 'concurrent:docs', 'compress']);
  
  // release tasks
  grunt.registerTask('releasePatch', ['test', 'bump-only:patch', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMinor', ['test', 'bump-only:minor', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMajor', ['test', 'bump-only:major', 'contributors', 'changelog', 'bump-commit']);
  
  // clean, test, generate docs (the CI task)
  grunt.registerTask('all', ['clean', 'test', 'docs']);
};


index.js
/*!
 *
 * Copyright (c) 2013 Sebastian Golasch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

'use strict';

// ext. libs
var Q = require('q');
var os = require('os');
var cp = require('child_process');
var appium = require('appium/lib/server/main');
var portscanner = require('portscanner');

/**
 * This module is a browser plugin for [DalekJS](//github.com/dalekjs/dalek).
 * It provides all a WebDriverServer & browser launcher for Safari on iOS.
 * 
 * At the moment this only works with the IPhone 
 *
 * The browser plugin can be installed with the following command:
 *
 * ```bash
 * $ npm install dalek-browser-ios --save-dev
 * ```
 *
 * You can use the browser plugin by adding a config option to the your Dalekfile
 *
 * ```javascript
 * "browsers": ["ios"]
 * ```
 *
 * Or you can tell Dalek that it should test in this browser via the command line:
 *
 * ```bash
 * $ dalek mytest.js -b ios
 * ```
 *
 * The Webdriver Server tries to open Port 9003 by default,
 * if this port is blocked, it tries to use a port between 9004 & 9093
 * You can specifiy a different port from within your [Dalekfile](/pages/config.html) like so:
 *
 * ```javascript
 * "browsers": {
 *   "ios": {
 *     "port": 5555 
 *   }
 * }
 * ```
 *
 * It is also possible to specify a range of ports:
 *
 * ```javascript
 * "browsers": {
 *   "ios": {
 *     "portRange": [6100, 6120] 
 *   }
 * }
 * ```
 *
 * If you would like to test on the IPad (IPhone) emulator, you can simply apply a snd. argument,
 * which defines the browser type:
 *
 * ```bash
 * $ dalek mytest.js -b ios:ipad
 * ```
 * 
 * @module DalekJS
 * @class IosDriver
 * @namespace Browser
 * @part iOS
 * @api
 */

var IosDriver = {

  /**
   * Verbose version of the browser name
   *
   * @property longName
   * @type string
   * @default Mobile Safari iOS
   */

  longName: 'Mobile Safari iOS (iPhone)',

  /**
   * Default port of the Appium WebDriverServer
   * The port may change, cause the port conflict resolution
   * tool might pick another one, if the default one is blocked
   *
   * @property port
   * @type integer
   * @default 4723
   */

  port: 4723,

  /**
   * WebHook port
   *
   * @property webhookPort
   * @type integer
   * @default 9003
   */

  webhookPort: 9003,

  /**
   * Default host of the Appium WebDriverServer
   * The host may be overridden with
   * a user configured value
   *
   * @property host
   * @type string
   * @default localhost
   */

  host: 'localhost',

  /**
   * Root path of the appium webdriver server
   *
   * @property path
   * @type string
   * @default /wd/hub
   */

  path: '/wd/hub',

  /**
   * Default desired capabilities that should be
   * transferred when the browser session gets requested
   *
   * @property desiredCapabilities
   * @type object
   */

  desiredCapabilities: {
    device: 'iPhone Emulator',
    name: 'Safari remote via WD',
    app: 'safari',
    version: '6.1',
    browserName: ''
  },

  /**
   * Driver defaults, what should the driver be able to access.
   *
   * @property driverDefaults
   * @type object
   */

  driverDefaults: {
    viewport: true,
    status: {
      os: {
        arch: os.arch(),
        version: os.release(),
        name: 'Mac OSX'
      }
    },
    sessionInfo: true
  },

  /**
   * Special arguments that are needed to invoke
   * appium. These are the defaults, they need to be modified later on 
   *
   * @property appiumArgs
   * @type object
   */

  appiumArgs: {
    app: null,
    ipa: null,
    quiet: true,
    udid: null,
    keepArtifacts: false,
    noSessionOverride: false,
    fullReset: false,
    noReset: false,
    launch: false,
    log: false,
    nativeInstrumentsLib: false,
    safari: false,
    forceIphone: false,
    forceIpad: false,
    orientation: null,
    useKeystore: false,
    address: '0.0.0.0',
    nodeconfig: null,
    port: null,
    webhook: null
  },

  /**
   * Different browser types (iPhone / iPad)
   *
   * @property browserTypes
   * @type object
   */

  browserTypes: {

    /**
     * IPad emulator
     *
     * @property ipad
     * @type object
     */

    ipad: {
      name: 'iPad'
    }

  },

  /**
   * Resolves the driver port
   *
   * @method getPort
   * @return {integer} port WebDriver server port
   */

  getPort: function () {
    return this.port;
  },

  /**
   * Resolves the maximum range for the driver port
   *
   * @method getMaxPort
   * @return {integer} port Max WebDriver server port range
   */

  getMaxPort: function () {
    return this.maxPort;
  },

  /**
   * Resolves the webhook port
   *
   * @method getWebhookPort
   * @return {integer} WebHook server port
   */

  getWebhookPort: function () {
    return this.webhookPort;
  },

  /**
   * Resolves the maximum range for the webhook port
   *
   * @method getWebhookPort
   * @return {integer} WebHook Max WebHook port
   */

  getMaxWebhookPort: function () {
    return this.maxWebhookPort;
  },

  /**
   * Returns the driver host
   *
   * @method getHost
   * @return {string} host WebDriver server hostname
   */

  getHost: function () {
    return this.host;
  },

  /**
   * Launches appium & corresponding emulator or device,
   * kicks off the portscanner
   *
   * @method launch
   * @param {object} configuration Browser configuration
   * @param {EventEmitter2} events EventEmitter (Reporter Emitter instance)
   * @param {Dalek.Internal.Config} config Dalek configuration class
   * @return {object} promise Browser promise
   */

  launch: function (configuration, events, config) {
    var deferred = Q.defer();

    // store injected configuration/log event handlers
    this.reporterEvents = events;
    this.configuration = configuration;
    this.config = config;

    // check if the user wants to run the iPad emulator
    if (configuration && configuration.type === 'ipad') {
      this.longName = this.longName.replace('iPhone', 'iPad');
      this.appiumArgs.forceIpad = true;
    }

    // check for a user set port
    var browsers = this.config.get('browsers');
    if (browsers && Array.isArray(browsers)) {
      browsers.forEach(this._checkUserDefinedPorts.bind(this));
    }

    // check if the current port is in use, if so, scan for free ports
    portscanner.findAPortNotInUse(this.getPort(), this.getMaxPort(), this.getHost(), this._checkPorts.bind(this, deferred));
    return deferred.promise;
  },

  /**
   * Kills the Appium Server process,
   * kills simulator processses
   * with a slight timeout to prevent 
   * appium from throwing errors
   * 
   * @method kill
   * @chainable
   */

  kill: function () {
    // kill appium servers
    this.appiumServer.webSocket.server.close();
    this.appiumServer.rest.listen().close();
    // slight timeout for process killing
    setTimeout(this._processes.bind(this, this._kill.bind(this)), 1000);
    return this;
  },

  /**
   * Kills the non blacklisted simulator processes & restores
   * the stderr handler
   *
   * @method _kill
   * @param {object|null} err Error or null
   * @param {array} result List of currently running simulator processes
   * @chainable
   * @private
   */

  _kill: function (err, result) {
    // kill simulator processes
    result.forEach(this._killProcess.bind(this));
    // (re)establish stderr/stdout stream
    this._reinstantiateLog();
    return this;
  },

  /**
   * Checks a blacklist & kills the process when
   * not found
   *
   * @method _killProcess
   * @param {integer} processID Process ID
   * @chainable
   * @private
   */

  _killProcess: function (processID) {
    var kill = true;

    // walk through the list of processes that are
    // open before the driver started
    this.openProcesses.forEach(function (pid) {
      if (pid === processID) {
        kill = false;
      }
    });

    if (kill === true) {
      cp.spawn('kill', [processID]);
    }

    return this;
  },

  /**
   * Checks & switches the appium server port,
   * scans the range for the webhook port
   *
   * @method _listProcesses
   * @param {object} deferred Promise
   * @param {object|null} err Error or null
   * @param {integer} port Appium server port to use
   * @chainable
   * @private
   */

  _checkPorts: function (deferred, error, port) {
    // check if the port was blocked & if we need to switch to another port
    if (this.port !== port) {
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to port: ' + port);
      this.port = port;
    }

    // check if the current webhook port is in use, if so, scan for free ports
    portscanner.findAPortNotInUse(this.getWebhookPort(), this.getMaxWebhookPort(), this.getHost(), this._launch.bind(this, deferred));
    return this;
  },

  /**
   * Checks & switches the webhook port,
   * loads a list of running simulator processes
   *
   * @method _listProcesses
   * @param {object} deferred Promise
   * @param {object|null} err Error or null
   * @param {integer} port Webhook port to use
   * @chainable
   * @private
   */

  _launch: function (deferred, error, port) {
    // check if the port was blocked & if we need to switch to another port
    if (this.webhookPort !== port) {
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to webhook port: ' + port);
      this.webhookPort = port;
    }

    // launch appium & the emulator
    this._processes(this._listProcesses.bind(this, deferred));
    return this;
  },

  /**
   * Stores open processes,
   * suppresses stdout logs,
   * starts appium
   *
   * @method _listProcesses
   * @param {object} deferred Promise
   * @param {object|null} err Error or null
   * @param {array} result List of currently running simulator processes
   * @chainable
   * @private
   */

  _listProcesses: function (deferred, err, result) {
    // save list of open emulator processes, before we launched it
    this.openProcesses = result;
    // nasty hack to surpress socket.io debug reports from appium
    this._suppressAppiumLogs();
    // run appium
    appium.run(this._loadAppiumArgs(this.appiumArgs), this._afterAppiumStarted.bind(this, deferred));
    return this;
  },

  /**
   * Stores the appium server reference,
   * restores the stdout logs
   *
   * @method _afterAppiumStarted
   * @param {object} deferred Promise
   * @param {object} appiumServer Appium server instance
   * @chainable
   * @private
   */

  _afterAppiumStarted: function (deferred, appiumServer) {
    this.appiumServer = appiumServer;
    deferred.resolve();
    return this;
  },

  /**
   * Configures appium
   *
   * @method _loadAppiumArgs
   * @param {object} appiumArgs Appium specific configuration
   * @return {object} Modified appium configuration
   * @private
   */

  _loadAppiumArgs: function (appiumArgs) {
    appiumArgs.port = this.getPort();
    appiumArgs.webhook = this.getHost() + ':' + this.getWebhookPort();
    return appiumArgs;
  },

  /**
   * Process user defined ports
   *
   * @method _checkUserDefinedPorts
   * @param {object} browser Browser configuration
   * @chainable
   * @private
   */

  _checkUserDefinedPorts: function (browser) {
    this._checkAppiumPorts(browser);
    this._checkWebhookPorts(browser);
    return this;
  },

  /**
   * Process user defined appium ports
   *
   * @method _checkAppiumPorts
   * @param {object} browser Browser configuration
   * @chainable
   * @private
   */

  _checkAppiumPorts: function (browser) {
    // check for a single defined port
    if (browser.ios && browser.ios.port) {
      this.port = parseInt(browser.ios.port, 10);
      this.maxPort = this.port + 90;
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to user defined port: ' + this.port);
    }

    // check for a port range
    if (browser.ios && browser.ios.portRange && browser.ios.portRange.length === 2) {
      this.port = parseInt(browser.ios.portRange[0], 10);
      this.maxPort = parseInt(browser.ios.portRange[1], 10);
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to user defined port(s): ' + this.port + ' -> ' + this.maxPort);
    }

    return this;
  },

  /**
   * Process user defined webhook ports
   *
   * @method _checkWebhookPorts
   * @param {object} browser Browser configuration
   * @chainable
   * @private
   */

  _checkWebhookPorts: function (browser) {
    // check for a single defined webhook port
    if (browser.ios && browser.ios.webhookPort) {
      this.webhookPort = parseInt(browser.ios.webhookPort, 10);
      this.maxWebhookPort = this.webhookPort + 90;
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to user defined webhook port: ' + this.webhookPort);
    }

    // check for a webhook port range
    if (browser.ios && browser.ios.webhookPortRange && browser.ios.webhookPortRange.length === 2) {
      this.webhookPort = parseInt(browser.ios.webhookPortRange[0], 10);
      this.maxWebhookPort = parseInt(browser.ios.webhookPortRange[1], 10);
      this.reporterEvents.emit('report:log:system', 'dalek-browser-ios: Switching to user defined webhook port(s): ' + this.webhookPort + ' -> ' + this.maxWebhookPort);
    }

    return this;
  },

  /**
   * Tracks running simulator processes
   *
   * @method _processes
   * @param {function} fn Callback
   * @chainable
   * @private
   */

  _processes: function (fn) {
    var cmd = ['ps -ax', '|', 'grep "iPhone Simulator.app"'];
    cp.exec(cmd.join(' '), this._transformProcesses.bind(this, fn));
    return this;
  },

  /**
   * Transforms the process list output into
   * a json structure
   * 
   * @method _transformProcesses
   * @param {function} fn Callback
   * @param {null|object} err Error if error, null if not
   * @param {string} stdout Terminal output
   * @chainable
   * @private
   */

  _transformProcesses: function(fn, err, stdout){
    var result = [];
    stdout.split('\n').forEach(this._scanProcess.bind(this, result));
    fn(err, result);
    return this;
  },

  /**
   * Scans and transforms the process list
   *
   * @method _scanProcess
   * @param {array} result Transformed result
   * @param {string} line Process list entry
   * @chainable
   * @private
   */

  _scanProcess: function (result, line){
    var data = line.split(' ');
    data = data.filter(this._filterProcessItem);

    if (data[1] === '??') {
      result.push(data[0]);
    }

    return this;
  },

  /**
   * Filters process list items
   *
   * @method _filterProcessItem
   * @param {string} item Process list entry
   * @return {bool|string} Process item or false
   * @private
   */

  _filterProcessItem: function (item) {
    if (item !== '') {
      return item;
    }

    return false;
  },

  /**
   * Overwrite default stdout & stderr handler
   * to suppress some appium logs 
   *
   * @method _suppressAppiumLogs
   * @chainable
   * @private
   */

  _suppressAppiumLogs: function () {
    // TODO: Check if the log level of appium can be set to 0
    var _supLogs = function (data) {
      if (data.search('6minfo') === -1 && data.search('33mwarn') === -1 && data.search('90mdebug') === -1) {
        this.oldWrite.bind(process.stdout)(data);
      }
    }.bind(this);

    // store old std. handler
    this.oldWrite = process.stdout.write;
    this.oldWriteErr = process.stderr.write;

    // overwrite with ugliness
    process.stdout.write = _supLogs;
    process.stderr.write = _supLogs;
    return this;
  },

  /**
   * Reinstantiate stdout handler after appium has
   * been started
   *
   * @method _reinstantiateLog
   * @chainable
   * @private
   */

  _reinstantiateLog: function () {
    setTimeout(function () {
      process.stdout.write = this.oldWrite;
      process.stderr.write = this.oldWriteErr;
    }.bind(this), 8000);
    return this;
  }

};

// expose the module
module.exports = IosDriver;
